### ISSUE TITLE
'no-unsafe-finally' rule gives errors on safe code

### ISSUE THREAD

**Issue by bakkot** [eslint/eslint#5972]
**What version of ESLint are you using?**
eslint@e606523

**What parser (default, Babel-ESLint, etc.) are you using?**
default

**What did you do? Please include the actual source code causing the issue.**
Rule `no-unsafe-finally` gives an error on the following code:
```js
try {} finally {
  while(true) {
    break;
  }
}
```

**What did you expect to happen?**
No errors.

**What actually happened? Please include the actual, raw output from ESLint.**
`Unsafe usage of BreakStatement                          no-unsafe-finally`

More generally, break and continue statements need to check that the statement being broken or continued is not interior to the finally block. This involves checking labels, if present.

ping @onurtemizkan, @platinumazure.

See #5808, #5932.

**Comment by bakkot** [comment]
And I guess if we're being pedantic, it's possible that this shouldn't give a warning either:

```js
try {} finally {
  try {
    throw 0;
  } catch (e) {
    //pass
  }
}
```

Of course, it's not possible to statically determine if a `throw` will cause an early exit from a `finally`:

```js
try {} finally {
  inner: try {
    throw 0;
  } finally {
    if (Math.random() < 0.5) {
      break inner;
    }
  }
}
```

But this is sufficiently horrifying as to be worth warning anyway. Frankly I'd be OK with warning in both of the cases mentioned here; it's more important that `break` and `continue` are handled correctly.

**Comment by louy** [comment]
Also what about this?
```js
try {} finally {
  (function() {
    return;
  })();
}
```

I'm not sure how to test it myself as this rule hasn't been released yet. The code above should give no errors.

**Comment by bakkot** [comment]
@louy, that one's handled [correctly](https://github.com/eslint/eslint/blob/e6065235a682c5beae4a3f6e05cd7fa1b2813a3e/tests/lib/rules/no-unsafe-finally.js#L25).

Easiest way to test it yourself is to clone master, `npm install`, then `./bin/eslint.js --no-eslintrc --rule 'no-unsafe-finally: 1' test.js`.

**Comment by onurtemizkan** [comment]
@louy this is ok. I think this issue is only about usages of `break` and `continue`

**Comment by onurtemizkan** [comment]
It's obviously a bug, I presume it will be accepted. So I'm now working on a patch to fix it asap.

**Comment by mysticatea** [comment]
Oh, good catch!
